File: experiment/bm25.rs
use std::collections::{HashMap, HashSet};
use std::f64::consts::LN_2;

fn compute_bm25(
    documents: &Vec<String>,
    query: &str,
    k1: f64,
    b: f64,
) -> Vec<f64> {
    let n = documents.len(); // Total number of documents
    let avg_doc_length: f64 =
        documents.iter().map(|doc| doc.split_whitespace().count()).sum::<usize>() as f64 / n as f64;

    let term_frequencies: Vec<HashMap<String, usize>> = documents.iter().map(|doc| {
        let mut tf = HashMap::new();
        for word in doc.split_whitespace() {
            *tf.entry(word.to_string()).or_insert(0) += 1;
        }
        tf
    }).collect();

    let mut document_frequencies = HashMap::new();
    for doc in documents.iter() {
        let unique_words: HashSet<_> = doc.split_whitespace().collect();
        for word in unique_words {
            *document_frequencies.entry(word.to_string()).or_insert(0) += 1;
        }
    }

    documents.iter().enumerate().map(|(index, doc)| {
        let doc_length = doc.split_whitespace().count();
        query.split_whitespace().fold(0.0, |mut score, term| {
            if let Some(&tf) = term_frequencies[index].get(term) {
                let df = *document_frequencies.get(term).unwrap_or(&0);
                let idf = ((n as f64 - df as f64 + 0.5) / (df as f64 + 0.5)).ln() + LN_2; // Adjusted IDF
                score += idf * ((tf as f64 * (k1 + 1.0)) /
                    (tf as f64 + k1 * (1.0 - b + b * (doc_length as f64 / avg_doc_length))));
            }
            score
        })
    }).collect()
}

fn main() {
    let documents = vec![
        "the quick brown fox".to_string(),
        "jumps over the lazy dog".to_string(),
        "the quick brown dog".to_string(),
        "quick".to_string(),
        "brown".to_string(),
        "browm".to_string(),
        "quick brown".to_string(),
    ];
    let query = "quick brown";
    let scores = compute_bm25(&documents, query, 1.5, 0.75);

    let mut mapped_documents: Vec<_> = documents.iter()
        .zip(scores.iter())
        .map(|(doc, &score)| (doc.clone(), score))
        .collect();

    mapped_documents.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());

    println!("Query \"{}\" BM25 Scores: {:?}", query, mapped_documents);
}

File: experiment/bm25.js
function computeBM25(documents, query, k1 = 1.5, b = 0.75) {
    const N = documents.length; // Total number of documents
    const avgDocLength =
        documents.reduce((sum, doc) => sum + doc.split(' ').length, 0) / N;

    const termFrequencies = documents.map((doc) => {
        const tf = {};
        doc.split(' ').forEach((word) => {
            tf[word] = (tf[word] || 0) + 1;
        });
        return tf;
    });

    const documentFrequencies = {};
    documents.forEach((doc) => {
        const uniqueWords = new Set(doc.split(' '));
        uniqueWords.forEach((word) => {
            documentFrequencies[word] = (documentFrequencies[word] || 0) + 1;
        });
    });

    const scores = documents.map((doc, index) => {
        const docLength = doc.split(' ').length;
        let score = 0;
        query.split(' ').forEach((term) => {
            if (term in termFrequencies[index]) {
                const df = documentFrequencies[term] || 0;
                const idf = Math.log((N - df + 0.5) / (df + 0.5) + 1);
                const tf = termFrequencies[index][term];
                score +=
                    idf *
                    ((tf * (k1 + 1)) /
                        (tf + k1 * (1 - b + b * (docLength / avgDocLength))));
            }
        });
        return score;
    });

    return scores;
}

// Example usage
function testBM25(documents, query, shouldJoin = false) {
    const scores = computeBM25(documents, query);
    const mappedDocuments = documents
        .map((doc, index) => [doc, scores[index]])
        .sort((a, b) => b[1] - a[1]);
    if (shouldJoin) {
        query = query.split(' ').join('');
        mappedDocuments.forEach((item) => {
            item[0] = item[0].split(' ').join('');
        });
    }

    console.log('Query', `"${query}"`, 'BM25 Scores:', mappedDocuments);
}

const documents = [
    'the quick brown fox',
    'jumps over the lazy dog',
    'the quick brown dog',
    'quick',
    'brown',
    'browm',
    'quick brown',
];
const query = 'quick brown';
testBM25(documents, query);

const documents1 = [
    'ខ្ញុំ ស្រលាញ់ អ្នក',
    'តាម ដែល អ្នក បាន ស្គាល់',
    'មិន អាច ស្គាល់ បាន ទេ',
    'ស្រលា',
    'អ្នក',
    'ខ្ញុំ',
];
const query1 = 'ខ្ញុំ ស្រលាញ់';
testBM25(documents1, query1);

const documents2 = [
    'ស្រលាស់',
    'ស្រលាញ់',
    'ស្លាញ់',
    'លាញ់',
    'លាញ',
    'រលាញ់',
    'ក្រលាញ់',
    'ឆ្នាំង',
    'មិន',
];
const query2 = 'រលាញ';
testBM25(
    documents2.map((item) => item.split('').join(' ')),
    query2.split('').join(' '),
    true,
);
